### ISSUE TITLE
Potential memory leak in certificate handling post 2.10 upgrade

### ISSUE THREAD

**Issue by radarnex** [containous/traefik#10213]
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.traefik.io) and didn't find any.

### What did you do?

Upgrade Traefik from `2.8.4` to `2.10.5`.

### What did you see instead?

After upgrading to `2.10.5`, we observed an unusual increase in memory usage suggestive of a memory leak. This issue surfaced a few hours post-upgrade and persists across various clusters.

![traefik_memory_leak_2 10](https://github.com/traefik/traefik/assets/15276213/bac209e4-cdc6-4b64-9d72-76eca738e8a4)

To diagnose, we enabled the debug endpoint, debug logs and collected profiling data. Also, we downgraded to `2.9.10` and the issue was happening as well.

Upon reviewing changes from `2.8.4` to `2.9.1`, only PR [ACME default certificate](https://github.com/traefik/traefik/pull/9189) seemed potentially relevant, but its connection to the issue isn't clear.

Other possibility we're thinking about is some bug/regression in the `crypto` library in Go since Go 1.20 (Traefik 2.9), whereas it wasn't present in Go 1.19 (Traefik 2.8).

Comparatively, the `AppendCertificate` function in `pkg/tls/certificate.go` is markedly more active in profiles from `2.9.1` or higher, whereas it was negligible in `2.8.4` profile.

### 2.10.5 (or 2.9)
<img width="635" alt="pprof_graph-2 10" src="https://github.com/traefik/traefik/assets/15276213/b43ab64a-acdf-4ca7-bef9-02844fcc7dc4">

### 2.8.4
<img width="738" alt="pprof_graph-2 8" src="https://github.com/traefik/traefik/assets/15276213/4907ff2c-767d-439f-a4d3-10c212403528">

I've attached the pprof top output for both versions below for reference. 

### 2.10.5
```
(pprof) top -sum
Active filters:
   ignore=sum
Showing nodes accounting for 151.01MB, 77.39% of 195.14MB total
Dropped 191 nodes (cum <= 0.98MB)
Showing top 10 nodes out of 208
      flat  flat%   sum%        cum   cum%
   51.07MB 26.17% 26.17%    51.07MB 26.17%  encoding/pem.Decode
   28.07MB 14.39% 40.56%    28.07MB 14.39%  bufio.NewReaderSize
   19.68MB 10.09% 50.64%    19.68MB 10.09%  bufio.NewWriterSize
   19.51MB 10.00% 60.64%    19.51MB 10.00%  math/big.nat.make (inline)
      16MB  8.20% 68.84%       16MB  8.20%  crypto/internal/bigmod.NewNat (inline)
    5.01MB  2.57% 71.40%     5.01MB  2.57%  regexp/syntax.(*compiler).inst
    3.51MB  1.80% 73.21%     3.51MB  1.80%  sigs.k8s.io/json/internal/golang/encoding/json.(*decodeState).literalStore
       3MB  1.54% 74.75%    94.58MB 48.47%  github.com/traefik/traefik/v2/pkg/tls.(*Certificate).AppendCertificate
    2.64MB  1.36% 76.10%     3.18MB  1.63%  compress/flate.NewWriter
    2.51MB  1.29% 77.39%     2.51MB  1.29%  golang.org/x/net/http2/hpack.(*headerFieldTable).addEntry
```

### 2.8.4
```
(pprof) top -sum
Active filters:
   ignore=sum
Showing nodes accounting for 59.18MB, 68.68% of 86.17MB total
Showing top 10 nodes out of 329
      flat  flat%   sum%        cum   cum%
   25.53MB 29.63% 29.63%    25.53MB 29.63%  bufio.NewWriterSize
   14.36MB 16.67% 46.30%    14.36MB 16.67%  bufio.NewReaderSize
    6.01MB  6.97% 53.27%     6.01MB  6.97%  regexp/syntax.(*compiler).inst
    2.51MB  2.91% 56.18%     2.51MB  2.91%  github.com/vulcand/oxy/utils.CopyHeaders
    2.50MB  2.90% 59.08%     2.50MB  2.90%  github.com/aws/aws-sdk-go/aws/endpoints.init
       2MB  2.32% 61.40%        2MB  2.32%  net.ParseCIDR
       2MB  2.32% 63.72%        2MB  2.32%  github.com/prometheus/client_golang/prometheus.MakeLabelPairs
    1.76MB  2.05% 65.77%     1.76MB  2.05%  compress/flate.NewWriter
    1.50MB  1.74% 67.51%     1.50MB  1.74%  github.com/json-iterator/go.(*Iterator).ReadString
    1.01MB  1.17% 68.68%     1.01MB  1.17%  encoding/base64.(*Encoding).DecodeString
```

### What version of Traefik are you using?

```
Version:      2.10.5
Codename:     saintmarcelin
Go version:   go1.21.3
Built:        2023-10-11T13:54:02Z
OS/Arch:      linux/amd64
```

### What is your environment & configuration?

**Kubernetes** (either 1.23 or 1.25).

**Traefik resources**
- Routers: ~650
- Services: ~110
- Middlewares: ~400
- TLS certificates: ~100

Args:
```
      --entrypoints.metrics.address=:9100/tcp
      --entrypoints.traefik.address=:9000/tcp
      --entrypoints.web.address=:8000/tcp
      --entrypoints.websecure.address=:8443/tcp
      --api.dashboard=true
      --ping=true
      --metrics.prometheus=true
      --metrics.prometheus.entrypoint=metrics
      --metrics.prometheus.addRoutersLabels=true
      --providers.kubernetescrd
      --providers.kubernetescrd.ingressClass=traefik
      --providers.kubernetescrd.allowCrossNamespace=true
      --providers.kubernetescrd.allowExternalNameServices=true
      --providers.kubernetescrd.allowEmptyServices=true
      --providers.kubernetesingress
      --providers.kubernetesingress.allowExternalNameServices=true
      --providers.kubernetesingress.allowEmptyServices=true
      --providers.kubernetesingress.ingressendpoint.publishedservice=ingress/traefik
      --providers.kubernetesingress.ingressClass=traefik
      --entrypoints.web.http.redirections.entryPoint.to=:443
      --entrypoints.web.http.redirections.entryPoint.scheme=https
      --entrypoints.web.proxyProtocol.trustedIPs=<cidr1>,<cidr2>
      --entrypoints.websecure.http.tls=true
      --entrypoints.websecure.forwardedHeaders.insecure
      --entrypoints.websecure.proxyProtocol.trustedIPs=<cidr1>,<cidr2>
      --log.format=json
      --accesslog=true
      --accesslog.format=json
      --accesslog.fields.defaultmode=keep
      --accesslog.fields.headers.defaultmode=drop
      --accesslog.fields.headers.names.Cf-Ipcountry=keep
      --accesslog.fields.headers.names.Cf-Ray=keep
      --accesslog.fields.headers.names.Referer=keep
      --accesslog.fields.headers.names.User-Agent=keep
      --accesslog.fields.headers.names.X-Forwarded-For=keep
```

### If applicable, please paste the log output in DEBUG level

Nothing relevant.

**Comment by ldez** [comment]
Hello,

to help diagnose the problem, can you try Traefik versions between v2.8.4 and v2.10.5 (minor and patch)?

**Comment by lbenguigui** [comment]
Hello @radarnex,

Thank you for your feedback!

We are still unable to reproduce the issue. Could you please try the following:
- increase the memory to ensure its a leak and not an increase of memory usage.
- if the leak is confirmed, create a pprof export when the memory usage is low and an another when it's high to be able to compare memory usage
- If possible, provide a reproducible use case.

**Comment by Geczy** [comment]
im using traefik 2.10.7 and it went from 1gb to 2gb overnight:

![CleanShot 2024-01-02 at 10 09 11@2x](https://github.com/traefik/traefik/assets/1036968/a72c8a75-c14b-42ec-8110-13aad17240ba)

**Comment by Geczy** [comment]
restarting the proxy, it immediately shoots up to 1gb

**Comment by Lucaber** [comment]
We are experiencing a similar issue.
Memory heap size starts at ~600MB and increases over time (on kubernetes ingress updates)


![image](https://github.com/traefik/traefik/assets/15067830/a87ddbc6-53ae-4c7e-b14a-986309152fee)
![image](https://github.com/traefik/traefik/assets/15067830/45f87576-3b38-4765-8d71-3ec851328d6b)

**Comment by Lucaber** [comment]
With a memory limit of 3GB, I am seeing OOMKills every ~3 days. Im trying to debug GarbageCollection with:
```
        - name: GOGC
          value: "50"
        - name: GODEBUG
          value: gctrace=1
```

After a single ingress update, i am seeing multiple garbage collection runs in the container logs. Traefik is running in a Kubernetes cluster with ca 3000 ingresses.

**Comment by nerzhul** [comment]
Hello, we may be faced to same issue where traefik is oomkiiled every day on our 400 nodes clusters.
We noticed this after disabling our HPA which hided the issue. We disabled mostly every feature we use but we kept kubernetes SD and isolated one ingress only in a single ingress class, to observe traefik over time.

Here you see a fresh 6Mb hop without any traffic added.

![image](https://github.com/traefik/traefik/assets/119752/22f92ebb-35fc-415e-8baf-7095bee5f46e)

Here is what we see onto our production clusters with 5k ingresses, 1 oomkiil per day (and a resync of the oomkill moment since yesterday where we added 1Gb to see)

![image](https://github.com/traefik/traefik/assets/119752/373a07a1-14cc-48bb-b02c-358c68fc856c)

**Comment by apamildner** [comment]
I'm not really seeing a memory leak pattern, but we do see an increase in memory usage by traefik after upgrade to `2.11.2`. We upgraded from `2.9.1`.  This made our HPA go crazy and scale all the pods. Seems like we got a "static" increase of 20mb on the pods after the update on average. We got about 50 ingress routes. I'm adding datadog memory metrics below.

*Explanation of image*
We updated at 09.00 in the morning, from `2.9.1` to `2.11.2` we can see the immediate static memory increase of about 20 mi. We did not get any significant change in traffic during this time. Subsequently we can see our HPA scaling to a lot of pods that were not really needed because we had heuristics of when to scaled based on memory, which was now wrong after the update
<img width="820" alt="image" src="https://github.com/traefik/traefik/assets/10640074/b04c4ca2-730d-472d-ab06-263aa9f375c8">
